plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts

```diff
@@ -0,0 +1,247 @@
+  context,
+  Span,
+  SpanAttributes,
+  SpanStatusCode,
+  trace,
+} from '@opentelemetry/api';
+import { getRPCMetadata, RPCType } from '@opentelemetry/core';
+import {
+  InstrumentationBase,
+  InstrumentationConfig,
+  InstrumentationNodeModuleDefinition,
+  safeExecuteInTheMiddle,
+} from '@opentelemetry/instrumentation';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import type { HookHandlerDoneFunction } from 'fastify/types/hooks';
+import type { FastifyInstance } from 'fastify/types/instance';
+import type { FastifyReply } from 'fastify/types/reply';
+import type { FastifyRequest } from 'fastify/types/request';
+import {
+  AttributeNames,
+  FastifyNames,
+  FastifyTypes,
+} from './enums/AttributeNames';
+import type { HandlerOriginal } from './types';
+import { safeExecuteInTheMiddleMaybePromise } from './utils';
+import { VERSION } from './version';
+
+export const ANONYMOUS_NAME = 'anonymous';
+
+/** Fastify instrumentation for OpenTelemetry */
+export class FastifyInstrumentation extends InstrumentationBase {
+  constructor(config: InstrumentationConfig = {}) {
+    super(
+      '@opentelemetry/instrumentation-fastify',
+      VERSION,
+      Object.assign({}, config)
+    );
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<any>(
+        'fastify',
+        ['^3.0.0'],
+        (moduleExports, moduleVersion) => {
+          this._diag.debug(`Applying patch for fastify@${moduleVersion}`);
+          return this._patchConstructor(moduleExports);
+        }
+      ),
+    ];
+  }
+
+  private _hookOnRequest() {
+    const instrumentation = this;
+    return function onRequest(
+      request: FastifyRequest,
+      reply: FastifyReply,
+      done: HookHandlerDoneFunction
+    ) {
+      if (!instrumentation.isEnabled()) {
+        return done();
+      }
+      const rpcMetadata = getRPCMetadata(context.active());
+      const routeName = request.routerPath;
+      if (routeName && rpcMetadata?.type === RPCType.HTTP) {
+        rpcMetadata.span.setAttribute(SemanticAttributes.HTTP_ROUTE, routeName);
+        rpcMetadata.span.updateName(`${request.method} ${routeName || '/'}`);
+      }
+      done();
+    };
+  }
+
+  private _wrapHandler(
+    pluginName: string,
+    hookName: string,
+    original: () => Promise<any> | void
+  ): () => Promise<any> | void {
+    const instrumentation = this;
+    return function (this: any, ...args): Promise<any> | void {
+      if (!instrumentation.isEnabled()) {
+        return original.apply(this, args);
+      }
+
+      const spanName = `${FastifyNames.MIDDLEWARE} - ${
+        original.name || ANONYMOUS_NAME
+      }`;
+
+      const span = instrumentation._startSpan(spanName, {
+        [AttributeNames.FASTIFY_TYPE]: FastifyTypes.MIDDLEWARE,
+        [AttributeNames.PLUGIN_NAME]: pluginName,
+        [AttributeNames.HOOK_NAME]: hookName,
+      });
+
+      return context.with(trace.setSpan(context.active(), span), () => {
```

Comment:
Setting the context here is good, but we also need to unset it upon calling the `done` callback, otherwise, context leaks from the middleware to the next middlewares and handlers.

for example:
```js
import fastify from 'fastify';
const app = fastify({logger: true});
app.addHook('onRequest', async (req, reply, done) => {
  // some logic
  done();
});
app.get('/:id', (req, reply) => {
  reply.send('hello world');
});
app.listen(3000);
```

This app produces a request handler which is the child of the middleware span instead of the HTTP span.
I suggest wrapping the `done` callback (the one we pass to the hook via `original.apply(this, args)`) with a wrapper that invokes it with the original context (as well as ending the span when the done callback is called)